pkg/admin/data/backup.go

```diff
@@ -0,0 +1,152 @@
+package data
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/openshift/microshift/pkg/config"
+	"github.com/openshift/microshift/pkg/util"
+
+	"k8s.io/klog/v2"
+)
+
+var (
+	cpArgs = []string{
+		"--verbose",
+		"--recursive",
+		"--preserve",
+		"--reflink=auto",
+	}
+)
+
+// MakeBackup backs up MicroShift data (/var/lib/microshift) to
+// target/name/ (e.g. /var/lib/microshift-backups/backup-00001).
+func makeBackup(cfg BackupConfig) error {
+	klog.InfoS("Backup started", "cfg", cfg)
+
+	if err := cfg.Validate(); err != nil {
+		return fmt.Errorf("invalid BackupConfig: %w", err)
+	}
+
+	if err := ensureDirExists(cfg.Storage); err != nil {
+		return err
+	}
+
+	dest := filepath.Join(cfg.Storage, cfg.Name)
+	dest_tmp := dest + ".tmp"
+	dest_old := dest + ".old"
+
+	if err := removePath(dest_tmp); err != nil {
+		return err
+	}
+
+	if err := copyDataDir(dest_tmp); err != nil {
+		return err
+	}
+
+	backupAlreadyExists, err := pathExists(dest)
+	if err != nil {
+		return err
+	} else if backupAlreadyExists {
+		klog.V(2).InfoS("Backup already exists - renaming temporarily", "path", dest)
+		if err := renamePath(dest, dest_old); err != nil {
+			return err
+		}
+	}
+
+	if err := renamePath(dest_tmp, dest); err != nil {
+		klog.Errorf("Renaming path failed - renaming %s back and deleting %s: %v", dest_old, dest_tmp, err)
```

Comment:
`dest_old` will only exist if the original location existed when this function started, so if the rename on line 60 fails for some other reason then the logic in this section is going to produce more errors.

What are the situations that could result in the location already existing? We can definitely write code to produce a unique name, so I think we only have to worry about the collision if the name comes from the user via the command line? What if we just refuse to take the backup in that case? Or don't let the user specify the full backup location (give us a directory and we pick a name in that directory)?

Reply:
Same code will be used in automated ostree backing up so 3rd boot into the same healthy deployment will overwrite backup taken during 2nd boot.

Reply:
We might want to move some of that overwrite logic "up" the call stack to a place where the code knows whether it is being invoked by the user or by automation. That will make the edge cases easier to reason about.

Reply:
Added function to interface `BackupExists(BackupName)` and it's called in the CLI. It will refuse to overwrite a backup unless `--force` is provided